### PR TITLE
fix: support webpack 5 FileSystemInfo API

### PR DESCRIPTION
**Pull request by vankop**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

https://github.com/webpack-contrib/stylelint-webpack-plugin/issues/187

### Breaking Changes

no

### Additional Info

no

### CODE REVIEW

File: src/LintDirtyModulesPlugin.js [hunk 1]
```diff
@@ -2,6 +2,25 @@ import { isMatch } from 'micromatch';
...
+function mapFileTimestampsToOldFormat(fileTimestamps) {
+  for (const [filename, FileSystemInfoEntry] of fileTimestamps.entries()) {
+    fileTimestamps.set(
+      filename,
+      FileSystemInfoEntry ? FileSystemInfoEntry.timestamp : null
+    );
+  }
+
+  return fileTimestamps;
+}
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: We have a big loss here mapping all the entries.
We can just do a simple check on `hasFileChanged`
No need this function

File: src/LintDirtyModulesPlugin.js [hunk 2]
```diff
@@ -2,6 +2,25 @@ import { isMatch } from 'micromatch';
...
+
+function getFileTimestamps(compilation) {
+  return (
+    compilation.fileTimestamps ||
+    // eslint-disable-next-line no-underscore-dangle
+    mapFileTimestampsToOldFormat(compilation.fileSystemInfo._fileTimestamps)
+  );
+}
+
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: No need this function, create a const in apply function

File: src/LintDirtyModulesPlugin.js [hunk 3]
```diff
@@ -14,16 +33,17 @@ export default class LintDirtyModulesPlugin {
   apply(compilation, callback) {
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: Create a const here to check if exists `fileSystemInfo`
```js
const fileTimestamps = compilation.fileTimestamps || compilation.fileSystemInfo._fileTimestamps
```

File: src/LintDirtyModulesPlugin.js [hunk 4]
```diff
@@ -14,16 +33,17 @@ export default class LintDirtyModulesPlugin {
...
     if (this.isFirstRun) {
       this.isFirstRun = false;
-      this.prevTimestamps = compilation.fileTimestamps;
+      this.prevTimestamps = getFileTimestamps(compilation);
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: Use new const `fileTimestamps`

File: src/LintDirtyModulesPlugin.js [hunk 5]
```diff
@@ -14,16 +33,17 @@ export default class LintDirtyModulesPlugin {
...
       callback();
       return;
     }
 
     const dirtyOptions = { ...this.options };
+    const newTimestamps = getFileTimestamps(compilation);
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: No need

File: src/LintDirtyModulesPlugin.js [hunk 6]
```diff
@@ -14,16 +33,17 @@ export default class LintDirtyModulesPlugin {
...
     const glob = dirtyOptions.files.join('|').replace(/\\/g, '/');
-    const changedFiles = this.getChangedFiles(compilation.fileTimestamps, glob);
+    const changedFiles = this.getChangedFiles(newTimestamps, glob);
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: Use new const `fileTimestamps`

File: src/LintDirtyModulesPlugin.js [hunk 7]
```diff
@@ -14,16 +33,17 @@ export default class LintDirtyModulesPlugin {
...
 
-    this.prevTimestamps = compilation.fileTimestamps;
+    this.prevTimestamps = newTimestamps;
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js]: Use new const `fileTimestamps`

**Comment by ricardogobbosouza** [comment]
We need to add tests

File: src/LintDirtyModulesPlugin.js [hunk 8]
```diff
@@ -2,6 +2,25 @@ import { isMatch } from 'micromatch';
...
+  }
+
+  return fileTimestamps;
+}
+
+function getFileTimestamps(compilation) {
+  return (
+    compilation.fileTimestamps ||
+    // eslint-disable-next-line no-underscore-dangle
+    mapFileTimestampsToOldFormat(compilation.fileSystemInfo._fileTimestamps)
```
Comment (sokra) [inline src/LintDirtyModulesPlugin.js]: `_fileTimestamps` is private. Don't do that! That's not how FileSystemInfo works. `fileSystemInfo._fileTimestamps` is a cache, while in webpack 4 `compilation.fileTimestamps` was a complete list.

Use the public functions to access timestamps of files.

Comment (ricardogobbosouza) [reply src/LintDirtyModulesPlugin.js]: Just use `getDeprecatedFileTimestamps` from FileSystemInfo

Comment (ricardogobbosouza) [reply src/LintDirtyModulesPlugin.js]: @sokra could we have a method in `getFileTimestamps` so as not to depend on a `getDeprecatedFileTimestamps` function that will be depreciated?

```js
getFileTimestamps() {
 return this._fileTimestamps;
}
```

File: src/LintDirtyModulesPlugin.js (hunk 1 above)
Comment (vankop) [reply src/LintDirtyModulesPlugin.js]: yep, you are right

**Comment by vankop** [comment]
Thanks for all your replies!

> `_fileTimestamps` is private. Don't do that! That's not how FileSystemInfo works. `fileSystemInfo._fileTimestamps` is a cache

I found out that it is cache (all timestamps that was created via `getFileTimestamp` method), but did not understand how to take all timestamps in other way (since `getDeprecatedFileTimestamps` also use cache) or my vision is mistaken and we must use just `getDeprecatedFileTimestamps` here? Maybe we can use snapshots here somehow? @sokra could you please guide me a little bit here

> Just use getDeprecatedFileTimestamps from FileSystemInfo

@ricardogobbosouza Unfortunately `getDeprecatedFileTimestamps` will not work as expected either. Hope @sokra will clarify usage a little bit.

> We need to add tests

So I can upgrade dev dependency on `webpack@5` ?

**Comment by vankop** [comment]
Lets wait for https://github.com/webpack/webpack/pull/9679

**Comment by thasmo** [comment]
@vankop, seems like the PR was merged and should be included with the next beta (beta.7) when released.

**Comment by lf-jeremy** [comment]
beta.7 is out and does include the PR

**Comment by vankop** [comment]
@evilebottnawi sorry I did not have time, hope on the nearest weekend I can try to migrate on beta.7 .

**Comment by ricardogobbosouza** [comment]
@vankop @evilebottnawi 
I created another PR #199

**Pull request closed without merging**